### ISSUE TITLE
Failed cluster sizing recommendations when DaemonSets do not run on all nodes

### ISSUE THREAD

**Issue by jcharcalla** [kubecost/cost-analyzer-helm-chart#1642]
**Describe the bug**
When DaemonSets do not run on all of the nodes in a cluster due to node pools, tolerations, taints, etc. The cluster sizing recommendation will fail with an error similar to this...

```error converting cluster cluster-one WorkloadDistribution to cluster sizing: failed to recommend single pool cluster sizing: failed to find machine pool for vCPUs: 1723.027448 (235.195698 DaemonSet); RAM: 4681.172393 GB (712.473709 DaemonSet); max pod: 13.000000 vCPUs, 40.000000 RAM GB with buffer 1.333333, min nodes 2```

This has been discussed previously in Zen desk tickets [1869](https://kubecost.zendesk.com/agent/tickets/1869), [2185](https://kubecost.zendesk.com/agent/tickets/2185) and [2422](https://kubecost.zendesk.com/agent/tickets/2422).

Paraphrasing @michaelmdresser comment in [ZD: 1869](https://kubecost.zendesk.com/agent/tickets/1869)...

In this case the [cluster sizing code](https://github.com/kubecost/kubecost-cost-model/blob/develop/pkg/savings/clustersizing.go#L81-L84) is looking for a node with `(13*1.3 + 235.2*1.3)` CPU cores and `(40*1.3 + 712.5*1.3)` RAM GB which cannot be satisfied by any [AWS node types](https://github.com/kubecost/kubecost-cost-model/blob/develop/pkg/provider/aws.go#L434).
 
**To Reproduce**
1. Create a cluster with 3 nodes, each with a unique taint
2. Create 3 DaemonSets, each with a toleration matching one of the node taints
3. Observe that the WorkloadDistribution generated during cluster sizing has higher-than-expected numbers for DaemonSet CPU and RAM

**Expected behavior**
Actual daemonSet usage is used in calculating the required CPU/Mem instead of assuming daemonSet pods run on each node in the cluster.

**Comment by dwbrown2** [comment]
Great report! @michaelmdresser this one seems like it could be hard or am I overestimating?

**Comment by michaelmdresser** [comment]
Yep, probably quite challenging. We have no concept of node pools/groups right now. Fixing this would require introducing that concept to the cluster right-sizing algorithm.

**Comment by Lax77** [comment]
We are also seeing the similar issue., in our case I did notice on one such cluster on one of the node Daemonset was unable to run the pod and it got evicted.

ERR savings: cluster sizing: error reccomending cluster sizing for cluster <our cluster name>: failed to recommend single pool cluster sizing: failed to find machine pool for vCPUs: 2476.461222 (194. │
│ 010140 DaemonSet); RAM: 23190.903387 GB (384.573252 DaemonSet); max pod: 5.000000 vCPUs, 48.000000 RAM GB with buffer 1.333333, min nodes

**Comment by Lax77** [comment]
We also noticed this issue when a daemonset is unable to run due to lack of resources(CPU, Memory) on a node.

**Comment by srpomeroy** [comment]
I have another customer experiencing the same issue.
```2022-09-30T15:18:48.866518294Z ERR savings: cluster sizing: error reccomending cluster sizing for cluster cluster-one: failed to recommend single pool cluster sizing: failed to find machine pool for vCPUs: 8449.137473 (551.990303 DaemonSet); RAM: 18554.122618 GB (647.945532 DaemonSet); max pod: 47.000000 vCPUs, 230.077000 RAM GB with buffer 1.333333, min nodes 2```

**Comment by Adam-Stack-PM** [comment]
@AjayTripathy, A few customers have reported. Could you review and see if you would like to escalate this issue for v1.00.  May be a challenging fix. 

CC @dwbrown2 for product input.

**Comment by AjayTripathy** [comment]
I actually think this can be done in v1.99. @biancaburtoiu has a list of bugs to fix on this page and I think we can find a day or two to knock them out.

**Comment by anoop2811** [comment]
Hi, wanted to check  if this going to be fixed in 1.99? I think of this as a fundamental feature since righsizing the nodes can make a difference in cost so not having this functionality is a big ding.

**Comment by MrJW27** [comment]
Hi @AjayTripathy / @biancaburtoiu  - have we confirmed that this will be resolved in 1.99?

**Comment by biancaburtoiu** [comment]
We aim to have a beta solution for this in 1.99. We'll share more details on the usage closer to the release. Thank you all for your patience!

**Comment by biancaburtoiu** [comment]
To address the issue of the current cluster sizing recommendations algorithm not taking into account node/workload-level constraints, eg. daemonsets not running on all nodes in the cluster, v1.99 will introduce a **new beta API endpoint** that determines the workloads of each node in the cluster individually to produce node-level sizing recommendations.

**Note 1**: This endpoint is **not** available in the Kubecost UI just yet but can be accessed at 
```go
<your-path>/model/savings/nodeSizingETL?window=2d&targetUtilization=0.75&allowSharedCore=false&cluster=cluster-one
```
**Note 2**: This endpoint is in a **beta** state. Please share your feedback with us!

The new endpoint accepts several request parameters:
- `cluster` (required, eg. `cluster-one`)
- `window` (optional, default `2d`)
- `spotNodes` (optional, default `false`)
- `targetUtilization` (optional, default `0.8`)
- `allowSharedCore` (optional, default `true`)
- `architecture` (optional, default `""` - unspecified)

The response should be interpreted as follows:
- `currentInformation.summary` contains a summary of the current cluster state: the available nodes, their resources (CPU/RAM) and estimated utilization
- `currentInformation.breakdown` contains the same information at individual node level
- for each node in `currentInformation.breakdown` there will exist an equivalent entry in `recommendations.breakdown` representing our computed recommendation: a new machine type, alongside its estimated utilization, monthly cost and subsequent savings (if applicable)
- `recommendations.summary` contains a summary of the individual node recommendations described in `recommendations.breakdown`, including a `monthlyRate` summation of node-level `monthlyRate`s and resource utilization averaged over all recommendation nodes

**Comment by mmurph3** [comment]
@biancaburtoiu per customer:

When we tried, we’re getting 404 on this endpoint <your-path>/model/savings/nodeSizingETL?window=2d&targetUtilization=0.75&allowSharedCore=false&cluster=cluster-one . We also tried without /model . And, since we were not sure what the following note in github issue meant, This endpoint is not available in the Kubecost UI just yet but can be accessed.. , we tried on using both paths. With kubecost-ui and the api server. With the kubecost-ui, we see that without /model , it’s a 200 response. But, it’s a blank page.

**Comment by srpomeroy** [comment]
@mmurph3 This works in our demo environment. Can you verify the customer has upgraded to 1.99?

https://demo.kubecost.xyz/model/savings/nodeSizingETL?window=2d&targetUtilization=0.75&allowSharedCore=false&cluster=kc-demo-prod
